### ISSUE TITLE
useQuery when throwing an error not working as per documentation

### ISSUE THREAD

**Issue by bmmpt** [tannerlinsley/react-query#2672]
**Describe the bug**
The documentation states the following (https://react-query.tanstack.com/guides/query-functions):

"Handling and Throwing Errors
For React Query to determine a query has errored, the query function must throw. Any error that is thrown in the query function will be persisted on the error state of the query."

But useQuery seems to be letting the exception bubble and the component "breaks".

**To Reproduce**
https://codesandbox.io/s/react-query-throw-not-working-jmmqf

**Expected behavior**
The component should render: <h1>Not WORKING!!!</h1>
The 3 console logs should be written to the console. `console.log(error, "error");` should white the "Query error" message thrown.

**Desktop (please complete the following information):**
 - Windows 10
 - Browser: Chrome
 - Version: Latest
 - React 17.0.2
 - React Query 3.23.2

I'm interpreting the documentation incorrectly?

If I downgrade to an older version (2.x) of React Query it works as per documentation.

**Comment by TkDodo** [comment]
you forgot to mention that you are using `suspense: true,`. Per default, `suspense` uses Error Boundaries to catch and handle errors. This is well documented on the [suspense docs page](https://react-query.tanstack.com/guides/suspense), together with how to opt-out of it, which is by setting `useErrorBoundary: false`.

**Comment by bmmpt** [comment]
@TkDodo = lifesaver :)

It makes total sense. I built a Framework that handles a lot of boiler plate concerns and it handles errors without Error Boundaries. I recently made a change to my http service hook to throw the error (which it was already handling), so that optimistic updates and deletes would run onError in case the http service call fails.
I recall reading something about the `useErrorBoundary` flag weeks back, but since it had no impact until my latest change, I didn't remember all that I had read, I did some research now, but it didn't lead me to that flag.
I will turn the flag off and it should be all good. But a curiosity question comes to mind... Why didn't I experience the same behavior when I use useMutation? I'm throwing the error inside my http service when I use both `useQuery` and `useMutation`, but my component doesn't 'break' on an 'errored' mutation...

I updated the sandbox to exemplify:
https://codesandbox.io/s/react-query-throw-not-working-jmmqf
